hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/payload/ExpressionPayload.scala

```diff
@@ -411,10 +414,14 @@ object ExpressionPayload {
     parseSchema(props.getProperty(PAYLOAD_RECORD_AVRO_SCHEMA))
   }
 
-  private def getWriterSchema(props: Properties): Schema = {
-    ValidationUtils.checkArgument(props.containsKey(HoodieWriteConfig.WRITE_SCHEMA_OVERRIDE.key),
-      s"Missing ${HoodieWriteConfig.WRITE_SCHEMA_OVERRIDE.key} property")
-    parseSchema(props.getProperty(HoodieWriteConfig.WRITE_SCHEMA_OVERRIDE.key))
+  private def getWriterSchema(props: Properties, isPartialUpdate: Boolean): Schema = {
+    if (isPartialUpdate) {
+      parseSchema(props.getProperty(HoodieWriteConfig.WRITE_PARTIAL_UPDATE_SCHEMA.key))
```

Comment:
I was actually expecting we always use the full schema as the write schema, while for partial update, it takes along another configuration for the partial schema, which is more straight forward.

Reply:
It depends on what the `SCHEMA` header in the log block refers to.  My assumption is that the `SCHEMA` header corresponds to the records written to the log block, so if the records are partial, `SCHMEA` is partial, and we use another header to indicate it's partial, to differentiate from schema evolution.

Either way, do you agree that we need to keep both full and partial schemas in the log block header?

Reply:
I'm not very approval with two schemas here, partial update is actually a special case of schema evolution (that only supplementing new columns and never drop columns), the writer should still takes the full schema always, and we can persist the partial schema to the lock block for some optimization purposes (for example, the handing of missing values as nulls, with partial schema, we can know that whether a null value comes from partial update or it should be force updated as null).

Reply:
Generally we may have 3 modes for fields that not updated in partial update:
1. keep it as it is;
2. force update it as null; (which I think should never happen in real case);
3. overwrite with default (if the detault is defined in the schema)

I think 1 is the most natural handling, but in any case, there reader should always use it's own reader schema for merging, not the writer schema.

Another question is when to evolve the table schema, does it happends before or after the commit succeed?

Reply:
Agree that option 1 is the most natural handling.

In the current schema evolution on write, the write schema is evolved based on the input, and the evolved schema is written to the commit.